File: 024treeIncludes.js
const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");
const g = new Node("g");
const h = new Node("h");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;
e.left = g;
f.right = h;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f
//    /       \
//   g         h

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.left = null;
//     this.right = null;
//   }
// }

const treeIncludes = (root, target) => {
  if (root === null) return false;
  const stack = [root];
  while (stack.length > 0) {
    const current = stack.shift();
    if (current.val === target) return true;
    if (current.left) stack.push(current.left);
    if (current.right) stack.push(current.right);
  }
  return false;
};

treeIncludes(a, "p"); // -> false

File: 024tree_includes.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

def tree_includes(root, target):
  if root is None:
    return False

  stack = [root]

  while stack:
    current = stack.pop()
    if current.val == target:
      return True

    if current.right is not None:
      stack.append(current.right)
      
    if current.left is not None:
      stack.append(current.left)

  return False
